bundler/lib/bundler/checksum.rb

```diff
@@ -2,12 +2,94 @@
 
 module Bundler
   class Checksum
+    class << self
+      def digests_from_file_source(file_source, digest_algorithms: %w[sha256])
+        raise ArgumentError, "not a valid file source: #{file_source}" unless file_source.respond_to?(:with_read_io)
+
+        digests = digest_algorithms.map do |digest_algorithm|
+          [digest_algorithm.to_s, Bundler::SharedHelpers.digest(digest_algorithm.upcase).new]
+        end.to_h
+
+        file_source.with_read_io do |io|
+          until io.eof?
+            block = io.read(16_384)
+            digests.each_value {|digest| digest << block }
+          end
+
+          io.rewind
+        end
+
+        digests
+      end
+
+      def from_lock(checksums, source)
+        checksums.split(",").map do |c|
+          algo, digest = c.split("-", 2)
+          new(algo, digest, source)
+        end
+      end
+
+      def to_lock(checksums)
+        checksums.map(&:to_lock).sort.join(",")
+      end
+
+      def match_digests?(checksums, digests)
+        return true if checksums.empty? && digests.empty?
+
+        common_algos = checksums.keys & digests.keys
+        return true if common_algos.empty?
+
+        common_algos.all? do |algo|
+          checksums[algo].digest == digests[algo]
+        end
+      end
+    end
+
+    attr_reader :algo, :digest, :sources
+    def initialize(algo, digest, source)
+      @algo = algo
+      @digest = digest
+      @sources = Set.new
+      @sources << source
+    end
+
+    def ==(other)
+      other.is_a?(self.class) && other.digest == digest && other.algo == algo && sources == other.sources
+    end
+
+    def hash
+      digest.hash
+    end
+
+    alias_method :eql?, :==
+
+    def to_s
+      "#{algo}-#{digest} (from #{sources.first}#{", ..." if sources.size > 1})"
+    end
+
+    def to_lock
+      "#{algo}-#{digest}"
+    end
+
+    def merge!(other)
+      raise ArgumentError, "cannot merge checksums of different algorithms" unless algo == other.algo
+      unless digest == other.digest
+        raise SecurityError, <<~MESSAGE
+          #{other}
+          #{self} from:
+          * #{sources.join("\n* ")}
+        MESSAGE
+      end
+      @sources.merge!(other.sources)
```

Comment:
```suggestion
      @sources.merge(other.sources)
```
This is what I get for working from my phone.